plugins/inputs/sqlserver/sqlserver.go

```diff
@@ -423,6 +464,96 @@ func (s *SQLServer) Init() error {
 	return nil
 }
 
+// Get Token Provider by loading cached token or refreshed token
+func getTokenProvider() (func() (string, error), error) {
+	var tokenString string
+
+	// load token
+	muCacheLock.RLock()
+	token, err := loadToken()
+	muCacheLock.RUnlock()
+
+	// if there's error while loading token or found an expired token, refresh token and save it
+	if err != nil || token.IsExpired() {
+		// refresh token within a write-lock
+		muCacheLock.Lock()
+
+		// load token again, in case it's been refreshed by another thread
+		token, err = loadToken()
+
+		// check loaded token's error/validity, then refresh/save token
+		if err != nil || token.IsExpired() {
+			// get new token
+			spt, err := refreshToken()
+			if err != nil {
+				return nil, err
+			}
+
+			// use the refreshed token
+			tokenString = spt.OAuthToken()
+		} else {
+			// use locally cached token
+			tokenString = token.OAuthToken()
+		}
+
+		muCacheLock.Unlock()
```

Comment:
Oh boy, if you return in between you will not release the lock!!!!!!!!!!!!!
Please use
```golang
muCacheLock.Lock()
defer muCacheLock.Unlock()
```
as a robust pattern wherever possible!

Reply:
Thank you for catching this, fixed it